torch/csrc/autograd/autograd_not_implemented_fallback.cpp

```diff
@@ -45,8 +45,170 @@ void _foreach_tensor(
   }
 }
 
+AutogradFallbackMode kAutogradFallbackMode = AutogradFallbackMode::Warn;
+
 } // namespace
 
+void setAutogradFallbackMode(AutogradFallbackMode mode) {
+  TORCH_CHECK(mode != AutogradFallbackMode::Error, "NYI: mode='error'");
+  kAutogradFallbackMode = mode;
+}
+
+AutogradFallbackMode getAutogradFallbackMode() {
+  return kAutogradFallbackMode;
+}
+
+static void warnAutogradNotImplemented(const std::string& op_name) {
+  TORCH_WARN(
+      op_name,
+      ": an autograd kernel was not registered to the Autograd key(s) ",
+      "but we are trying to backprop through it. This may lead to silently incorrect behavior. ",
+      "This behavior is deprecated and will be removed in a future version of PyTorch. ",
+      "If your operator is differentiable, please ensure you have registered an "
+      "autograd kernel to the correct Autograd key (e.g. DispatchKey::Autograd, "
+      "DispatchKey::CompositeImplicitAutograd). If your operator is not "
+      "differentiable, or to squash this warning and use the previous behavior, "
+      "please register torch::CppFunction::makeFallthrough() to DispatchKey::Autograd.");
```

Comment:
Is it worth specifically pointing out interactions with Python key? For example if I trace with LoggingMode, that can result in the autograd key being excluded and changing the results of my computation.

Reply:
>  For example if I trace with LoggingMode, that can result in the autograd key being excluded and changing the results of my computation.

What do you mean? I'm not familiar with how LoggingMode looks like anymore

Reply:
```python
from torch.library import Library
from torch.testing._internal.logging_tensor import capture_logs_with_logging_tensor_mode

test_ns = "abc"
lib = Library(test_ns, "FRAGMENT")
lib.define("foo(Tensor(a!) a) -> Tensor(a!)")

def get_op(name):
    return getattr(getattr(torch.ops, test_ns), name).default

op = get_op("foo")

# User initially defines this as a composite op
def foo_impl(a):
    a *= 2
    return a

lib.impl("foo", foo_impl, "CPU")
x = torch.randn(3, requires_grad=True)
x_clone = x.clone()

with capture_logs_with_logging_tensor_mode() as logs:
  y = op(x_clone)

dx, = torch.autograd.grad(y.sum(), x)
print(dx)
```
Normally, autograd key is a fallthrough, so when we run foo_impl everything is OK, we build the graph as expected and the gradient is correct `tensor([2., 2., 2.])`
But, if logging_mode is enabled, then dx returns a tensor full of ones because foo_impl would be run with autograd key disabled.

Reply:
I see, thanks for the explanation.

I'd prefer to skip an explanation of the interaction with the Python dispatch key to keep the error message simple; also,  I agree that the above behavior falls under "this may lead to silently incorrect behavior"

Reply:
sounds good!